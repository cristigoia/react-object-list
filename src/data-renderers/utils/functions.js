/**
 * Gets the value of the cell from the row and keys provided
 * @param  {Object} row  Row supplied
 * @param  {Array|string} keys List of keys used to access the value
 * @return {(Object|string|number)}      Value from row, or row itself if no keys
 */
const getValueFromAccessor = (row, keys) => {
  if (!Array.isArray(keys)) keys = keys.split('.')
  let value = row
  const cloneKeys = [...keys]
  while (value && cloneKeys.length) {
    const thisKey = cloneKeys.shift()
    if (thisKey in value) {
      value = value[thisKey]
    }
  }
  return value
}

export {
  getValueFromAccessor,
}
