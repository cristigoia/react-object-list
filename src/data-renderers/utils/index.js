export {
  getValueFromAccessor,
  handleRowClick,
} from './functions'
