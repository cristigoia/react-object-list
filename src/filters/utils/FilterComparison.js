import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

class FilterComparison extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: Select.propTypes.value,
    })),
    value: Select.propTypes.value,
    onChange: PropTypes.func.isRequired,
  }

  render() {
    const {options, value} = this.props
    return (
      <div className="objectlist-current-filter__filter-comparison">
        <Select
          options={options}
          value={value}
          onChange={newValue => this.props.onChange(newValue.value)}
          clearable={false}
        />
      </div>
    )
  }
}

export default FilterComparison
