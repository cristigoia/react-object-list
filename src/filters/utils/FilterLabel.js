import React from 'react'
import PropTypes from 'prop-types'

/**
 * Component to render label for filters
 */
class FilterLabel extends React.Component {
  static propTypes = {
    /** Text to render in the label */
    label: PropTypes.string.isRequired,
  }
  render() {
    return (
      <label className="objectlist-current-filter__filter-name">{this.props.label}:</label>
    )
  }
}

export default FilterLabel
