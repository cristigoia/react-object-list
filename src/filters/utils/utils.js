const possibleQueryStringLookupExpressions = ['exact', 'lt', 'gt', 'gte', 'lte']

export {
  possibleQueryStringLookupExpressions,
}
