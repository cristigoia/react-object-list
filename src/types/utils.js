function floatPrecision(floatValue, precision) {
  floatValue = parseFloat(floatValue)
  if (isNaN(floatValue)) { return parseFloat('0').toFixed(precision) } else {
    const power = Math.pow(10, precision)
    floatValue = (Math.round(floatValue * power) / power).toFixed(precision)
    return floatValue.toString()
  }
}

function commaNumber(numberString) {
  const parts = numberString.toString().split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',')
  return parts.join('.')
}
/**
 * Converts a float to stringed currency representation
 * @param  {number} floatNumber Amount to convert
 * @return {string}             Dollar string
 */
export function currency(floatNumber) {
  return `${floatNumber < 0 ? '-' : ''}$${commaNumber(floatPrecision(Math.abs(floatNumber), 2))}`
}
