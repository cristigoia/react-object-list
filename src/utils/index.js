export * from './constants'
export * from './proptypes'
export * from './functions'
export loadingSpinner from './loading'
export ErrorMessage from './ErrorMessage'
