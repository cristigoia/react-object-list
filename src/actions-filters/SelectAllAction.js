import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { loadingSpinner } from '../utils'

class SelectAllAction extends Component {
  static propTypes = {
    /** the loading status of the select all text */
    loading: PropTypes.bool,
    /** the total number of items in the dataset */
    count: PropTypes.number,
    /** the amount of items displayed on the current page */
    itemCount: PropTypes.number,
    /** the number of items currently selected */
    numSelected: PropTypes.number,
    /** callback function to handle selecting all items in the dataset. Set to null to not show it. */
    selectAll: PropTypes.func,
    /** callback function to handle deselecting all items */
    deselectAll: PropTypes.func,
  }
  static defaultProps = {
    loading: false,
    selectAllLink: true,
  }
  handleSelectAllClick = (event) => {
    event.preventDefault()
    this.props.selectAll()
  }
  handleDeselectAllClick = event => {
    event.preventDefault()
    this.props.deselectAll()
  }
  render() {
    let selectAll
    if (this.props.selectAll) {
      if (this.props.loading) {
        selectAll = (
          <a className="objectlist-button objectlist-button--link" href="#">
            {loadingSpinner()} Selecting all items ...
          </a>
        )
      } else if (this.props.count > 0 && this.props.numSelected < this.props.count && this.props.numSelected >= this.props.itemCount) {
        selectAll = (
          <a className="objectlist-button objectlist-button--link" href="#" onClick={this.handleSelectAllClick}>
            Select all {this.props.count.toLocaleString()} items
          </a>
        )
      }
    }
    let deselect
    if (this.props.numSelected > 0) {
      deselect = (
        <a className="objectlist-button objectlist-button--link" href="#" onClick={this.handleDeselectAllClick}>De-select all</a>
      )
    }
    return <div>{selectAll}{deselect}</div>
  }
}

export default SelectAllAction
