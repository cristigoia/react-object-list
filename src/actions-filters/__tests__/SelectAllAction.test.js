import React from 'react'
import { shallow } from 'enzyme'
import { snapshotTest } from 'utils/tests'
import SelectAllAction from '../SelectAllAction'

jest.mock('../../utils', () => ({
  loadingSpinner: () => 'loadingSpinner',
}))

describe('<SelectAllAction />', () => {
  describe('Snapshot', () => {
    it('does not render with no props', () => {
      snapshotTest(<SelectAllAction />)
    })
    it('shows both select all and deselect', () => {
      const props = {
        selectAll: jest.fn(),
        loading: false,
        count: 5,
        numSelected: 3,
        itemCount: 3,
      }
      snapshotTest(<SelectAllAction {...props} />)
    })
    it('shows deselect all when some of the items are selected', () => {
      const props = {
        loading: false,
        count: 5,
        numSelected: 2,
        itemCount: 3,
      }
      snapshotTest(<SelectAllAction {...props} />)
    })
    it('shows deselect all when all the items are selected', () => {
      const props = {
        loading: false,
        count: 5,
        numSelected: 5,
        itemCount: 3,
      }
      snapshotTest(<SelectAllAction {...props} />)
    })
    it('shows a spinner when loading', () => {
      const props = {
        selectAll: jest.fn(),
        loading: true,
        count: 5,
        numSelected: 3,
        itemCount: 3,
      }
      snapshotTest(<SelectAllAction {...props} />)
    })
    it('shows global select all when selected all on page', () => {
      const props = {
        selectAll: jest.fn(),
        loading: false,
        count: 5,
        numSelected: 3,
        itemCount: 3,
      }
      snapshotTest(<SelectAllAction {...props} />)
    })
  })
  describe('Functions', () => {
    let instance
    const baseProps = {}
    const mockEvent = {}
    beforeEach(() => {
      mockEvent['preventDefault'] = jasmine.createSpy()
      baseProps['selectAll'] = jasmine.createSpy()
      baseProps['deselectAll'] = jasmine.createSpy()
      instance = shallow(<SelectAllAction {...baseProps} />).instance()
    })
    it('selects all', () => {
      instance.handleSelectAllClick(mockEvent)
      expect(mockEvent.preventDefault).toHaveBeenCalled()
      expect(baseProps.selectAll).toHaveBeenCalled()
    })
    it('deselects all', () => {
      instance.handleDeselectAllClick(mockEvent)
      expect(mockEvent.preventDefault).toHaveBeenCalled()
      expect(baseProps.deselectAll).toHaveBeenCalled()
    })
  })
})
