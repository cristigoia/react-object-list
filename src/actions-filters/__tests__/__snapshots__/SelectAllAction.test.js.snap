// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectAllAction /> Snapshot does not render with no props 1`] = `null`;

exports[`<SelectAllAction /> Snapshot shows both select all and deselect 1`] = `
<div
  className="objectlist-row"
>
  <a
    className="objectlist-link"
    href="#"
    onClick={[Function]}
  >
    Select all 
    5
     
  </a>
  <a
    className="objectlist-link"
    href="#"
    onClick={[Function]}
  >
    De-select all
  </a>
</div>
`;

exports[`<SelectAllAction /> Snapshot shows deselect all when all the items are selected 1`] = `
<div
  className="objectlist-row"
>
  <a
    className="objectlist-link"
    href="#"
    onClick={[Function]}
  >
    De-select all
  </a>
</div>
`;

exports[`<SelectAllAction /> Snapshot shows deselect all when some of the items are selected 1`] = `
<div
  className="objectlist-row"
>
  <a
    className="objectlist-link"
    href="#"
    onClick={[Function]}
  >
    De-select all
  </a>
</div>
`;

exports[`<SelectAllAction /> Snapshot shows global select all when selected all on page 1`] = `
<div
  className="objectlist-row"
>
  <a
    className="objectlist-link"
    href="#"
    onClick={[Function]}
  >
    Select all 
    5
     
    kitties
  </a>
  <a
    className="objectlist-link"
    href="#"
    onClick={[Function]}
  >
    De-select all
  </a>
</div>
`;
